### ISSUE TITLE
Interactive pages formatting issues

### ISSUE THREAD

**Issue by KimPatch** [DukeLivingHistory/RLH-Site-Code#66]
**Checklist:** _For complex issues, you can create a checklist of tasks by using the following syntax:_

- [ ] Regular pages formatting interface is missing for interactive pages
- [ ] Format no-media deletes HTML tags
- [ ] Need to also check if HTML tags are read correctly after #64 is fixed
- [ ] Blog interactive pages also need to retain the formatting interface

**Other details:**   
I added some HTML tags to the white list to see if no-media formatting would ignore them that way. Ultimately HTML tags should be ignored along with the white space around them, and they should work just like they do in regular pages.

**Comment by ethanbutler** [comment]
@KimPatch – I have some serious concerns about this from a conceptual standpoint. WebVTT is not a markup language. It's designed specifically for text-based transcripts. The effort needing for parsing HTML would be akin to writing a new language parser.

What it seems like what you're looking for is the ability to link directly to individual sentences in an HTML document. The approach I'd advocate for would be either generate these links programmatically, or (if you want manual control) to use HTML data attributes. Something like.

```html
<p><span data-link>This is a linkable text group.</span><span data-link>This is another. These could be more than one sentence.</span></p>
```

Since we're using WordPress, we could even use a shortcode to make that a little easier for content creators.

The timestamp convention is a clever workaround for purely text-based content, but in order to support markup it's an extraneous level of abstraction that makes the implementation more complicated than it would be to implement a specifically-tailored HTML convention from scratch.

**Comment by KimPatch** [comment]
Understood.  (Just to be clear – the way everything works on the interview pages with the "format no media" button is great, and that needs to stay intact.)

But for pages and blogs, which also need HTML formatting, I see what you are saying. So let's use the short code [segment]. So what I'm picturing now is a "Segment for sharing" button that appears as part of/near the formatting interface that appears on pages and blogs. When the user presses the button it puts the segment short code above every sentence. We need the content creator to be able to adjust. Like the Format no-media button on the interview pages, there's a white list the user can add to.

So…

> [segment] 
> Here's a sentence of text. 
> [segment] 
> Here's a sentence of text.
> 
> [segment] 
> <ul>
> <li>As idiot proof as possible 
> <li>did not preclude keeping existing habits
> <li>The Living History site shows the InSite publishing system in action
> </ul>
> 
> [segment] 
> For details see our three-part report:<br>
> Part 1: The Interview: Journalism Cornerstone – and Roadblock<br>
> Part 2: InSite: a Guide<br>
> Part 3: Technology to Watch<br>

Is there anything I'm not thinking about?

**Comment by ethanbutler** [comment]
Another option that I hadn't considered would be using Markdown style formatting, like in GitHub comments. Something like:

```markdown
00:00:00.000 --> 00:00:01.000
# Heading

00:00:01.000 --> 00:00:02.000
Regular text **bold text** *italic text*

00:00:01.000 --> 00:00:02.000
* List
* List
* List
```

This would be a more complicated built, but it might be a better UX for content creators. Let me know your thoughts and I can provide an estimate for effort via email.

**Comment by KimPatch** [comment]
Answering my own question, there are two more things to think about:
**First**, this doesn't take supporting content into consideration.  

But it does seem like the WordPress short tag [segment] solution would fulfill the goal of having a method that anyone could use as a simple addition to a WordPress site that would provide direct links to content. 
Question: Could this be packaged as a WordPress plug-in?

So it looks like with the above plan the content creator would have a choice when putting a blog entry together:
- put an interactive page into the blog feed, but the text formatting would be limited.
- or use [segment] on a regular blog page but not be able to use supporting content 
This would also split where blog entries would appear on the backend, which is a minor minus.

**Second** – it looks like WebVTT does support the HTML tags we want. 
They work as expected if I put them into a Duke interview – except for one glitch – subsequent <v name> tags stop showing. If this could be fixed, then I think we just want to make sure the HTML tags don't get eaten when the content creator clicks Format. This might also be related to #62 

The editing buttons don't have to show up on the interactive pages, it's just that the Format no-media button shouldn't swallow the tags.

For lists, because there's a nested tag we might need another rule about how to handle them – I'm thinking we treat the whole list as a single sentence. That might be easiest. In that case, we might be able to take care of <li> just by putting it on the white list. 
Also <br> should be treated as a sentence.

Actually, maybe this can be solved just by adding a second list alongside the whitelist that content creators can add to – an additional end-of-sentence sentence indicator list.

Thoughts?

**Comment by KimPatch** [comment]
It looks like we were working on comments at the same time. I like markdown for its readability, but I'm hesitant to use multiple formatting standards across the site. The other question is does WebVTT support markdown?

**Comment by KimPatch** [comment]
This might help. http://rlh.demo.sqnts.xyz/interactives/the-insite-system/
This is a published interactive page that contains a bunch of HTML. I just had to add the HTML tags in after formatting automatically because they were deleted. So what I'm doing is formatting using the editing system and then copying to here and hitting the Format no media button. Again, would be even better if I could do formatting this window, but it will work if the tags just aren't deleted and we can  put the appropriate tags in an the end of sentence list.

**Comment by ethanbutler** [comment]
@KimPatch Addressing the above point-by-point:

- The shortcode functionality could totally be packaged as a WordPress plugin.
- WebVTT's support for tags isn't the issue here. The issue is the ability to write a parser than can reliably split HTML into sentences, which isn't possible, hence the need for shortcodes.
- WebVTT in this context is an abstraction layer that isn't useful. The site's ability to link to timestamps has nothing inherently to do with WebVTT. The linking functionality builds on top of WebVTT, but it could just as easily build on top of a different data structure that's more appropriate for HTML (like shortcodes). WebVTT is a video transcript format, and by adhering to it for different contexts, we're creating a potentially massive amount of technical debt for no real purpose. Think of WebVTT as being a "flat" format. We need something "nested."
- An approach we could take for nested structures like lists would be to have an option to indicate that a "link" is a continuation of the previous, so that when rendered child nodes could share the same identifier. Something like:

```
<ul>
  <li>[[linked-text]]First list item[[linked-text]]</li>
  <li>[[linked-text]]Second list item[[linked-text]]</li>
  <li>[[linked-text continued]]Third list item[[linked-text]]</li>
</ul>
```

Then when someone copies a link, they'd grab the same URL for both the second and third of those items.

**Comment by KimPatch** [comment]
I see why you suggested markdown – short codes aren't going to be very readable, and would require a lot of manual work. I think we get #78 working as well as possible, and maybe revisit this at some later point.

**Comment by KimPatch** [comment]
A question, harking back to the two unchecked items at the top:
Now that #78 is good, is it possible to put the editing interface on interactive pages or the Format No Media button on blog pages?

**Comment by ethanbutler** [comment]
No, for the same reasons:

- Interweaving HTML and this system is a can of worms that we’ve only scratched the surface of. I’ve been explicit about this. It might work in limited cases, like <a> tags, but it will fall apart at scale, as we’ve seen. By adding a WYSIWYG interface to UI, we’d be suggesting to users that this is a fully-supported feature. This implementation would be impossible to fully support.
- From a technical standpoint, having a content type (i.e. Posts) that supports “regular” content and this type of content becomes messy, because of some architectural aspects. This type of content is rendered client-side, whereas “normal” content is rendered server-side, and the logic for splitting that is a significant hurdle to get through. 

For these reasons, the idea of shortcodes was proposed. It handles the first point because shortcodes work fine with nested content. It handles the second point that we can bundle this functionality as a plugin that works independently of this site’s unique rendering paradigm.

We can iterate a bit over how these shortcodes look (as I know you didn’t like the sample above) but this is going to be the only way to move forward in this area.

**Comment by KimPatch** [comment]
Got it - thanks for the details. It helps to be able to picture why. Thanks again for your patience on this one. This is low priority for now, but we might want to eventually revisit your Multimarkdown idea as a Wordpress plugin.

**Comment by KimPatch** [comment]
Note: this is also related to #43 #63 and #78
